pkg/network/ovn_kubernetes_test.go

```diff
@@ -2148,6 +2149,220 @@ func TestRenderOVNKubernetesDualStackPrecedenceOverUpgrade(t *testing.T) {
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_HOST, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_CONTAINERIZED, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC, util.OVN_NAMESPACE, true, true, errMsg)
+
+	// move to phase 2 of the upgrade: ovnk master, ovnk node, ovnk control plane are all up
+	// The 4.13 ipsec daemonset should still be rendered along with the 4.14 ones.
+	fmt.Println("\n============= phase 1 is done, phase 2 is about to start; will start roll out of ovnkube node")
+	errMsg = "end of phase 1, will start phase 2"
+
+	bootstrapResult.OVN.NodeUpdateStatus.Progressing = false
+	bootstrapResult.OVN.MasterUpdateStatus.Progressing = false
+
+	objs, _, err = renderOVNKubernetes(config, bootstrapResult, manifestDirOvn, fakeClient, featureGatesCNO)
+	if err != nil {
+		t.Errorf("[%s] Unexpected error: %v", errMsg, err)
+	}
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_HOST, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_CONTAINERIZED, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC, util.OVN_NAMESPACE, true, true, errMsg)
+
+	// move to phase 2 of the upgrade: ovnk master, ovnk node, ovnk control plane are all up
+	// The 4.13 ipsec daemonset should still be rendered along with the 4.14 ones.
+	fmt.Println("\n============= phase 2 is ongoing: roll out of ovnk control plane")
+	errMsg = "phase 2 ongoing"
+
+	bootstrapResult.OVN.ControlPlaneUpdateStatus = &bootstrap.OVNUpdateStatus{
+		Kind:         "Deployment",
+		Namespace:    util.OVN_NAMESPACE,
+		Name:         util.OVN_CONTROL_PLANE,
+		Version:      "4.14.0",
+		IPFamilyMode: names.IPFamilySingleStack,
+		Progressing:  true,
+	}
+
+	bootstrapResult.OVN.NodeUpdateStatus.InterConnectZoneMode = "multizone" // ovnkube-node roll out is done and it's now on multizone
+
+	objs, _, err = renderOVNKubernetes(config, bootstrapResult, manifestDirOvn, fakeClient, featureGatesCNO)
+	if err != nil {
+		t.Errorf("[%s] Unexpected error: %v", errMsg, err)
+	}
+
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_HOST, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC_CONTAINERIZED, util.OVN_NAMESPACE, true, true, errMsg)
+	checkDaemonsetIsRendered(objs, util.OVN_IPSEC, util.OVN_NAMESPACE, true, true, errMsg)
+
+	// move to the end of the upgrade: ovnk master is gone, ovnk node, ovnk control plane are up.
+	// The 4.13 ipsec daemonset is gone; the two 4.14 ipsec daemonsets are finally rendered without
+	// the create-wait annotation.
+	fmt.Println("\n============= end of phase 2, will remove ovnkube master")
+	errMsg = "end of phase 2"
+
+	bootstrapResult.OVN.ControlPlaneUpdateStatus.Progressing = false // roll out of control plane is done
+	bootstrapResult.OVN.IPsecUpdateStatus.Name = util.OVN_IPSEC_HOST // ovn-ipsec-host now replaced ovn-ipsec
+	bootstrapResult.OVN.IPsecUpdateStatus.Version = "4.14.0"
```

Comment:
one question: i guess so far ovn-ipsec-host is rendered with create-wait-annotation ? so how it suddenly pops up from `IPsecUpdateStatus` ?

Reply:
It's related to the answer to your other question, my bad for not making it clear from the start. In the line you highlighted I'm setting the name of the ipsec update status to `ovn-ipsec-host`, while until then it was `ovn-ipsec`. This reflects what the bootstrap code in CNO would store in  `IPsecUpdateStatus`, but it won't make any difference in the running code, since the name of `IPsecUpdateStatus` is never checked. It's just to remind us what changes are supposed to be happening.